mikeio1d/pandas_extension/result_frame_aggregator.py

```diff
@@ -0,0 +1,332 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from typing import Callable
+
+if TYPE_CHECKING:
+    from typing import Any
+    from typing import List
+    from typing import Dict
+
+from dataclasses import fields
+
+import pandas as pd
+
+from mikeio1d.pandas_extension import groupby_level
+from mikeio1d.pandas_extension import compact_dataframe
+from mikeio1d.various import make_list_if_not_iterable
+from mikeio1d.quantities import TimeSeriesId
+
+
+class ResultFrameAggregator:
+    """
+    Aggregates a MIKE IO 1D result DataFrame into scalar values associate with entities.
+
+    Parameters
+    ----------
+    agg : str or callable
+        Default aggregation strategy for all levels.
+    override_name : str
+        Set a custom name for the overall aggregation.
+    **kwargs
+        Aggregation strategies for specific levels (e.g. time='min', chainage='mean')
+
+    Attributes
+    ----------
+    entity_levels : list of str
+        Entity levels are the geometric entities that the DataFrames are ultimately grouped by.
+    quantity_levels: list or str
+        Quantity levels are the levels which uniquely identify a quantity.
+    agg_levels : list of str
+        Agg levels are the levels that are aggregated along.
+    agg_strategies : dict of str: callable
+        Agg strategies are the strategies used for aggregation along the agg_levels.
+
+    Examples
+    --------
+    # Aggregate along chainage and time, taking the max of each quantity
+    >>> agg = ResultFrameAggregator('max')
+    >>> agg.aggregate(df)
+
+    # Always take the first chainage value, but take the max of time
+    >>> agg = ResultFrameAggregator('max', chainage='first')
+    >>> agg.aggregate(df)
+
+    # Same result as above, but with explicit argument names.
+    >>> agg = ResultFrameAggregator(time='max', chainage='first')
+    """
+
+    def __init__(self, agg: str | Callable = None, override_name: str = None, **kwargs):
+        kwargs.setdefault("time", agg)
+        if kwargs["time"] is None:
+            raise ValueError("Must specify an aggregation strategy for time.")
+        if agg is None:
+            agg = kwargs["time"]
+
+        self._override_name = override_name
+        self._agg_level_name = "agg"
+
+        self._entity_levels = ("group", "name", "tag")
+        self._agg_levels = ("duplicate", "chainage", "time")
+        self._quantity_levels = ("quantity", "derived")
+        self._agg_strategies = self._init_agg_strategies(agg, kwargs)
```

Comment:
I suggest to call the method _create_agg_strategies. For me _init is more for creating an instance variable and potentially not returning :-).